internal/graph/check_test.go

```diff
@@ -119,3 +121,201 @@ func TestCheckWithOneConcurrentGoroutineCausesNoDeadlock(t *testing.T) {
+			maxDBReads: 4,
+		},
+		{
+			name:       "intersection",
+			check:      tuple.NewTupleKey("document:x", "intersection", "user:maria"),
+			allowed:    true,
+			minDBReads: 2, // need at minimum two direct tuple checks
+			maxDBReads: 2,
+		},
+		{
+			name:       "intersection_no_access",
+			check:      tuple.NewTupleKey("document:x", "intersection", "user:unknown"),
+			allowed:    false,
+			minDBReads: 2, // need at minimum two direct tuple checks
+			maxDBReads: 4, // at most two tuple checks + two userset checks
+		},
+		{
+			name:       "difference",
+			check:      tuple.NewTupleKey("document:x", "difference", "user:jon"),
+			allowed:    true,
+			minDBReads: 2, // need at minimum two direct tuple checks
+			maxDBReads: 3, // at most two tuple checks + two userset checks
+		},
+		{
+			name:       "difference_no_access",
+			check:      tuple.NewTupleKey("document:x", "difference", "user:maria"),
+			allowed:    false,
+			minDBReads: 1, // if the "but not" condition returns quickly with "false", no need to evaluate the first branch
+			maxDBReads: 4, // at most two tuple checks + two userset checks
+		},
+		{
+			name:       "ttu",
+			check:      tuple.NewTupleKey("document:x", "ttu", "user:maria"),
+			allowed:    true,
+			minDBReads: 2, // one read to find org:fga + one direct check if user:maria is a member of org:fga
+			maxDBReads: 2,
+		},
+		{
+			name:       "ttu_no_access",
+			check:      tuple.NewTupleKey("document:x", "ttu", "user:jon"),
+			allowed:    false,
+			minDBReads: 3, // one read to find org:fga + (one direct check + userset check) to see if user:jon is a member of org:fga
+			maxDBReads: 3,
+		},
+		// more complex scenarios
+		{
+			name:       "union_and_ttu",
+			check:      tuple.NewTupleKey("document:x", "union_and_ttu", "user:maria"),
+			allowed:    true,
+			minDBReads: 3, // union (1 read) + ttu (2 reads)
+			maxDBReads: 3,
+		},
+		{
+			name:       "union_and_ttu_no_access",
+			check:      tuple.NewTupleKey("document:x", "union_and_ttu", "user:unknown"),
+			allowed:    false,
+			minDBReads: 3, // min(union (4 reads), ttu (3 reads)) = 3
+			maxDBReads: 4, // max(union (4 reads), ttu (3 reads)) = 4
+		},
+		{
+			name:       "union_or_ttu",
+			check:      tuple.NewTupleKey("document:x", "union_or_ttu", "user:maria"),
+			allowed:    true,
+			minDBReads: 1, // union path (1 read)
+			maxDBReads: 2, // ttu path (2 reads)
+		},
+		{
+			name:       "union_or_ttu_no_access",
+			check:      tuple.NewTupleKey("document:x", "union_or_ttu", "user:unknown"),
+			allowed:    false,
+			minDBReads: 11, // union (4 reads) + ttu (3 reads) + union rewrite (4 reads)
+			maxDBReads: 11,
+		},
+		{
+			name:       "intersection_of_ttus",
+			check:      tuple.NewTupleKey("document:x", "intersection_of_ttus", "user:maria"),
+			allowed:    true,
+			minDBReads: 4,
+			maxDBReads: 5,
+		},
+	}
+
+	// run the test many times to exercise all the possible DBReads
+	for i := 1; i < 1000; i++ {
```

Comment:
1000 iterations is a lot. It'll lead to slower tests.

5 seems reasonable to me?

Reply:
The entire test takes 6 seconds to run. I'd rather spend this time and have the test give me confidence

Reply:
Why does 1000 iterations give you any more confidence than 5? It's still random concurrent behavior. The probability doesn't change with an increasing number of iterations. Concurrency oddness is just as likely to happen on iteration 1 as it is iteration 1000.

Reply:
Doesn't need to block, but it would be nice to lower it 👍

Reply:
My thinking was that 5 iterations isn't enough to produce all the possible values of the number of db reads